ui,tracing: introduce react-based tracing UI

dhartunian opened the pull request:
Last two commits are in this PR 

![Screenshot 2022-02-28 at 10-38-38 Cockroach Console](https://user-images.githubusercontent.com/986307/156012231-766d3bd0-171e-45a7-8355-843be99a7909.png)


----
**tracing: add gRPC endpoints for tracez UI**

Previously, we had custom HTTP endpoints for the tracez UI.
This commit implements those endpoints on the Admin server
and defines the requests and responses in protobufs.

TODO(davidh): refactor `DoStuff` method
TODO(davidh): consider adding recording mode output to spans

Release note: None

----
**ui: implement active traces viewer in advanced debug**

This commit adds a single-page component that facilitates
viewing active trances and taking snapshots.

The state that this page manages is entirely self-contained
and does not interact with the application's redux store.

TODO(davidh): Add client-side routing so that URLs can be
shared.
TODO(davidh): Clarify user interface around setting trace
recording mode.
TODO(davidh): Tag display needs some iteration as some traces
can acquire very long tag lists that make the table extremely
wide.

Release note (ui change): Active operations can now be
snapshotted and inspected in a new page linked from
"Advanced Debug".

andreimatei committed:
tracing: fix verbose tag

The "_verbose" tag was inadvertently reported for a span if whoever is
requesting the span's recording was asking for a verbose recording.
However, it's possible to ask for a verbose recording from a span that's
not recording verbosely, in which case the tag should not be applied.
This patch makes the tag be applied depending on the span's recording
type, not based on what type of recording is requested.

Release note: None

andreimatei committed:
wip

andreimatei committed:
wip

andreimatei committed:
tracing: small cleanup

An inner function is no longer serving any purpose.

Release note: None

andreimatei committed:
kvserver/concurrency,tracing: add tracing tags while waiting for locks

When the evaluation of a request blocks on a lock acquisition, this
patch annotates the tracing span representing the evaluation with
tags about the lock's key, holder, time when the waiting started, and
number of locks that the evaluation waitied on previously.

These tags are a bit unusual in that they're temporary - they're removed
once the respective lock is released. As such, the patch adds a bit of
support for them to the tracing library. This API might evolve in the
future, as I play with more examples of spans that want to have a
"status".

Release note: None

andreimatei committed:
contention caption

andreimatei committed:
javascript monkeypatch

andreimatei committed:
propagation up and down

andreimatei committed:
styling

andreimatei committed:
propagate statement up

andreimatei committed:
toggle recording

andreimatei committed:
wip

andreimatei committed:
util/tracing: small refactor

Move a proto enum to the package scope, so it can be used more broadly
in the next patch.

Release note: None

andreimatei committed:
toggle recording

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/75685)
<!-- Reviewable:end -->

dhartunian commented:
@andreimatei didn't look through my work so not sure yet how big a mess I made on the server-side but see what you think of this so far! Roughly same number of lines as the HTML template and a good chunk of the functionality is there.

dhartunian committed:
tracing: add gRPC endpoints for tracez UI

Previously, we had custom HTTP endpoints for the tracez UI.
This commit implements those endpoints on the Admin server
and defines the requests and responses in protobufs.

TODO(davidh): refactor `DoStuff` method
TODO(davidh): consider adding recording mode output to spans

Release note: None

dhartunian committed:
ui: implement active traces viewer in advanced debug

This commit adds a single-page component that facilitates
viewing active trances and taking snapshots.

The state that this page manages is entirely self-contained
and does not interact with the application's redux store.

TODO(davidh): Add client-side routing so that URLs can be
shared.
TODO(davidh): Clarify user interface around setting trace
recording mode.
TODO(davidh): Tag display needs some iteration as some traces
can acquire very long tag lists that make the table extremely
wide.

Release note (ui change): Active operations can now be
snapshotted and inspected in a new page linked from
"Advanced Debug".

dhartunian commented:
@andreimatei updated this draft PR with two commits implementing backend and frontend for tracez. still needs some cleanup and some of the features are a little clunky but most of it is there.

dhartunian commented:
Closed in favor of #77712
